File: practice/leetcode/cpp/78_subsets.cpp
//https://leetcode.com/problems/subsets/description/
//Given an integer array nums of unique elements, return all possible
//subset (the power set).
//The solution set must not contain duplicate subsets. Return the solution in any order.

#include <vector>
using namespace std;

class Solutions {
public:
	vector<vector<int>> subsets(vector<int> &nums) {
		vector<vector<int>> answer;
		vector<int> row;
		
		for (int i = 0; i < (1 << nums.size()); i++) {
			int j = i;
			int indexcnt = 0;
			while(j) {
				if (j & 1) {
					row.push_back(nums[indexcnt]);
				}
				indexcnt++;
				j >>= 1;
			}
			row.clear();
			answer.push_back(row);
		}
		return(answer);
	} 
};

File: practice/leetcode/python/78_subsets.py
#https://leetcode.com/problems/subsets/description/
#Given an integer array nums of unique elements, return all possible
#subset (the power set).
#The solution set must not contain duplicate subsets. Return the solution in any order.

from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans, N = List[List[int]], len(nums)
        outer_iter = 2**N - 1

        while outer_iter:
            copy, i, subset = outer_iter, N - 1, List[int]
            while copy:
                if copy & 1:
                    subset.append(nums[i])
                i, copy = i - 1, copy >> 1

            ans.append(subset)
            outer_iter -= 1

        return(ans)
        

